src/main/kotlin/org/pkl/lsp/PklTextDocumentService.kt

```diff
@@ -28,58 +28,57 @@ import org.pkl.lsp.services.Topic
 
 val textDocumentTopic = Topic<TextDocumentEvent>("TextDocumentEvent")
 
-data class TextDocumentEvent(val file: URI, val type: TextDocumentEventType)
+sealed interface TextDocumentEvent {
+  val file: URI
 
-enum class TextDocumentEventType {
-  OPENED,
-  CHANGED,
-  CLOSED,
-  SAVED,
+  data class Opened(override val file: URI) : TextDocumentEvent
+
+  data class Closed(override val file: URI) : TextDocumentEvent
+
+  data class Saved(override val file: URI) : TextDocumentEvent
+
+  data class Changed(override val file: URI, val changes: List<TextDocumentContentChangeEvent>) :
+    TextDocumentEvent
```

Comment:
I wanted to include more details based on the event type, so, using an ADT here is a lot nicer. I changed all the events to follow the ADT pattern.